lib/ukstore/store.c

```diff
@@ -0,0 +1,1214 @@
+		int ret;
+		char to_set[UK_STORE_NR_MAX_LEN];
+
+		sprintf(to_set, "%" __PRIs64, val);
+		ret = e->set.charp(e->cookie, to_set);
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+int
+_uk_store_set_uptr(const struct uk_store_entry *e, __uptr val)
+{
+	UK_ASSERT(e);
+	switch (e->type) {
+	SETCASE_DOWNCASTUS(e, s8, S8, val, uptr);
+	SETCASE_DOWNCASTUU(e, u8, U8, val, uptr);
+	SETCASE_DOWNCASTUS(e, s16, S16, val, uptr);
+	SETCASE_DOWNCASTUU(e, u16, U16, val, uptr);
+	SETCASE_DOWNCASTUS(e, s32, S32, val, uptr);
+	SETCASE_DOWNCASTUU(e, u32, U32, val, uptr);
+	SETCASE_DOWNCASTUS(e, s64, S64, val, uptr);
+	SETCASE_UPCAST(e, u64, U64, val, uptr);
+	SETCASE_UPCAST(e, uptr, PTR, val, uptr);
+
+	case uk_store_get_entry_type(charp): {
+		int ret;
+		char to_set[UK_STORE_NR_MAX_LEN];
+
+		sprintf(to_set, "%p", (void *) val);
+		ret = e->set.charp(e->cookie, to_set);
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+int
+_uk_store_set_charp(const struct uk_store_entry *e, const char *val)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	case uk_store_get_entry_type(u8): {
+		__u8 to_set;
+
+		ret = sscanf(val, "%" __SCNu8, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.u8(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(s8): {
+		__s8 to_set;
+
+		ret = sscanf(val, "%" __SCNs8, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.s8(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(u16): {
+		__u16 to_set;
+
+		ret = sscanf(val, "%" __SCNu16, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.u16(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(s16): {
+		__s16 to_set;
+
+		ret = sscanf(val, "%" __SCNs16, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.s16(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(u32): {
+		__u32 to_set;
+
+		ret = sscanf(val, "%" __SCNu32, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.u32(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(s32): {
+		__s32 to_set;
+
+		ret = sscanf(val, "%" __SCNs32, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.s32(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(u64): {
+		__u64 to_set;
+
+		ret = sscanf(val, "%" __SCNu64, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.u64(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(s64): {
+		__s64 to_set;
+
+		ret = sscanf(val, "%" __SCNs64, &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.s64(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(uptr): {
+		__uptr to_set;
+
+		ret = sscanf(val, "%p", (void **) &to_set);
+		if (ret < 0)
+			return ret;
+
+		return e->set.uptr(e->cookie, to_set);
+	}
+
+	case uk_store_get_entry_type(charp): {
+		char *to_set = malloc(sizeof(*val) * strlen(val));
+
+		if (unlikely(!to_set))
+			return -ENOMEM;
+
+		strcpy(to_set, val);
+
+		ret = e->set.charp(e->cookie, to_set);
+
+		free(to_set);
```

Comment:
This makes sense, but doesn't feel right, we need to discuss how to manage memory inside it.
Most probably we need to pass to it also a buffer, or inform people that their buffer will be saved as a reference, not reallocated.